src/http.rs

```diff
@@ -0,0 +1,125 @@
+use hyper::mime::Mime;
+
+use registry::Registry;
+use errors::{Result, Error};
+use encoder::{Encoder, TextEncoder, TEXT_FORMAT};
+
+// run_with_registry runs a http server with a Registry, it blocks current thread.
+pub fn run_with_registry(addr: &str, reg: &Registry) -> Result<()> {
+    run(addr, reg, &TextEncoder {})
+}
+
+// run runs a http server with a Registry and a Encoder, it blocks current thread.
+pub fn run<'a>(addr: &str, reg: &Registry, encoder: &'a Encoder) -> Result<()> {
+    let reg = reg.clone();
+
+    let addr = try!(addr.parse().or_else(|e| Err(Error::Msg(format!("{:?}", e)))));
+    let server = try!(Server::http(&addr).or_else(|e| Err(Error::Msg(format!("{:?}", e)))));
+    if let Ok((_, server)) = server.handle(|_| HttpHandler::new(reg.clone(), encoder)) {
+        server.run();
+    }
+
+    Err(Error::Msg("http server error".to_owned()))
+}
+
+struct Buffer(Vec<u8>);
+
+impl Write for Buffer {
+    fn write(&mut self, v: &[u8]) -> io::Result<usize> {
+        self.0.extend_from_slice(v);
+        Ok(v.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+pub struct HttpHandler<'a> {
+    registry: Registry,
+    encoder: &'a (Encoder + 'a),
+    buffer: Buffer,
+    write_pos: usize,
+}
+
+impl<'a> HttpHandler<'a> {
+    pub fn new(registry: Registry, encoder: &'a Encoder) -> HttpHandler<'a> {
+        HttpHandler {
+            registry: registry,
+            encoder: encoder,
+            buffer: Buffer(Vec::new()),
+            write_pos: 0,
+        }
+    }
+}
+
+impl<'a> Handler<HttpStream> for HttpHandler<'a> {
+    fn on_request(&mut self, _: Request<HttpStream>) -> Next {
+        // TODO: route requests
+        Next::write()
+    }
+
+    fn on_request_readable(&mut self, _: &mut Decoder<HttpStream>) -> Next {
+        Next::write()
+    }
+
+    fn on_response(&mut self, res: &mut Response) -> Next {
+        if let Ok(written) = scarp(&self.registry, &mut self.buffer, self.encoder) {
+            res.headers_mut().set(ContentLength(written as u64));
+        } else {
+            return Next::remove();
+        }
+
+        self.write_pos = 0;
+        res.set_status(StatusCode::Ok);
+        res.headers_mut().set(ContentType(TEXT_FORMAT.parse::<Mime>().unwrap()));
+        Next::write()
+    }
+
+    fn on_response_writable(&mut self, encoder: &mut HyperEncoder<HttpStream>) -> Next {
+        match encoder.try_write(&self.buffer.0[self.write_pos..]) {
+            Ok(Some(n)) => {
+                if n == self.buffer.0.len() {
+                    Next::end()
+                } else {
+                    // a partial write
+                    self.write_pos = n;
+                    Next::write()
+                }
+            }
+            Ok(None) => Next::write(),
+            Err(_) => Next::remove(),
+        }
+    }
+}
+
+fn scarp(registry: &Registry, writer: &mut Write, encoder: &Encoder) -> Result<usize> {
```

Comment:
scrap is frequently used in [doc](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/writing_clientlibs.md#overall-structure)